File: src/leetcode/java/MaximumProductSubarray_152.java
package leetcode.java;

public class MaximumProductSubarray_152 {
  public int maxProduct(int[] nums) {
    int r = nums[0];
    int tmp;
    for (int i = 1, tmax = r, tmin = r; i < nums.length; i++) {
      if (nums[i] < 0) {
        tmp = tmax;
        tmax = tmin;
        tmin = tmp;
      }
      tmax = Math.max(nums[i], tmax * nums[i]);
      tmin = Math.min(nums[i], tmin * nums[i]);
      r = Math.max(r, tmax);
    }
    return r;
  }
}

File: src/leetcode/python/MaximumProductSubarray_152.py
class MaximumProductSubarray_152(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        r = nums[0]
        tmax = tmin = r
        for i in range(1, len(nums)):
            if nums[i] < 0:
                tmax, tmin = tmin, tmax
            tmax = max(nums[i], nums[i]*tmax)
            tmin = min(nums[i], nums[i]*tmin)
            r = max(tmax, r)
        return r
